File: applications/luci-app-upnp-igdv1/luasrc/model/cbi/upnp/upnp.lua
-- Copyright 2008 Steven Barth <steven@midlink.org>
-- Copyright 2008-2011 Jo-Philipp Wich <jow@openwrt.org>
-- Licensed to the public under the Apache License 2.0.

m = Map("upnpd", luci.util.pcdata(translate("Universal Plug & Play")),
	translate("UPnP allows clients in the local network to automatically configure the router."))

m:section(SimpleSection).template  = "upnp_status"

s = m:section(NamedSection, "config", "upnpd", translate("MiniUPnP settings"))
s.addremove = false
s:tab("general",  translate("General Settings"))
s:tab("advanced", translate("Advanced Settings"))

e = s:taboption("general", Flag, "enabled", translate("Enabled"))
e.rmempty  = false

s:taboption("general", Flag, "enable_upnp", translate("Enable UPnP functionality")).default = "1"
s:taboption("general", Flag, "enable_natpmp", translate("Enable NAT-PMP functionality")).default = "1"

s:taboption("general", Flag, "secure_mode", translate("Enable secure mode"),
	translate("Allow adding forwards only to requesting ip addresses")).default = "1"

s:taboption("general", Flag, "ipv6_disable", translate("Disable IPv6")).default = "1"

s:taboption("general", Flag, "ext_ip_reserved_ignore", translate("Ignore reserved ext_ip"),
	translate("This make the port forwarding force to work even when the router is behind NAT.")).default = "1"

s:taboption("general", Flag, "log_output", translate("Enable additional logging"),
	translate("Puts extra debugging information into the system log"))

s:taboption("general", Value, "download", translate("Downlink"),
	translate("Value in KByte/s, informational only")).rmempty = true

s:taboption("general", Value, "upload", translate("Uplink"),
	translate("Value in KByte/s, informational only")).rmempty = true

port = s:taboption("general", Value, "port", translate("Port"))
port.datatype = "port"
port.default  = 5000


s:taboption("advanced", Flag, "system_uptime", translate("Report system instead of daemon uptime")).default = "1"

s:taboption("advanced", Value, "uuid", translate("Device UUID"))
s:taboption("advanced", Value, "serial_number", translate("Announced serial number"))
s:taboption("advanced", Value, "model_number", translate("Announced model number"))

ni = s:taboption("advanced", Value, "notify_interval", translate("Notify interval (second)"))
ni.datatype    = "uinteger"
ni.placeholder = 30

ct = s:taboption("advanced", Value, "clean_ruleset_threshold", translate("Clean rules threshold (second)"))
ct.datatype    = "uinteger"
ct.placeholder = 20

ci = s:taboption("advanced", Value, "clean_ruleset_interval", translate("Clean rules interval (second)"))
ci.datatype    = "uinteger"
ci.placeholder = 600

pu = s:taboption("advanced", Value, "presentation_url", translate("Presentation URL"))
pu.placeholder = "http://192.168.9.1/"

s:taboption("advanced", Flag, "use_stun", translate("Use STUN"))

o = s:taboption("advanced", Value, "stun_host", translate("STUN host"))
o:depends("use_stun", "1")
o.datatype    = "host"
o:value("stun.qq.com")
o:value("stun.ekiga.net")
o:value("stun.syncthing.net")
o:value("stun.zoiper.com")
o:value("stun.gmx.net")
o.default = "stun.qq.com"

o = s:taboption("advanced", Value, "stun_port", translate("STUN port"))
o:depends("use_stun", "1")
o.datatype    = "port"
o:value("3478")
o.default = "3478"

lf = s:taboption("advanced", Value, "upnp_lease_file", translate("UPnP lease file"))
lf.placeholder = "/var/run/miniupnpd.leases"


s2 = m:section(TypedSection, "perm_rule", translate("MiniUPnP ACLs"),
	translate("ACLs specify which external ports may be redirected to which internal addresses and ports"))

s2.template  = "cbi/tblsection"
s2.sortable  = true
s2.anonymous = true
s2.addremove = true

s2:option(Value, "comment", translate("Comment"))

ep = s2:option(Value, "ext_ports", translate("External ports"))
ep.datatype    = "portrange"
ep.placeholder = "0-65535"

ia = s2:option(Value, "int_addr", translate("Internal addresses"))
ia.datatype    = "ip4addr"
ia.placeholder = "0.0.0.0/0"

ip = s2:option(Value, "int_ports", translate("Internal ports"))
ip.datatype    = "portrange"
ip.placeholder = "0-65535"

ac = s2:option(ListValue, "action", translate("Action"))
ac:value("allow", translate("Allow"))
ac:value("deny", translate("Deny"))

return m

File: applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
'use strict';
'require view';
'require dom';
'require poll';
'require uci';
'require rpc';
'require form';

var callInitAction, callUpnpGetStatus, callUpnpDeleteRule, handleDelRule;

callInitAction = rpc.declare({
	object: 'luci',
	method: 'setInitAction',
	params: [ 'name', 'action' ],
	expect: { result: false }
});

callUpnpGetStatus = rpc.declare({
	object: 'luci.upnp',
	method: 'get_status',
	expect: {  }
});

callUpnpDeleteRule = rpc.declare({
	object: 'luci.upnp',
	method: 'delete_rule',
	params: [ 'token' ],
	expect: { result : "OK" },
});

handleDelRule = function(num, ev) {
	dom.parent(ev.currentTarget, '.tr').style.opacity = 0.5;
	ev.currentTarget.classList.add('spinning');
	ev.currentTarget.disabled = true;
	ev.currentTarget.blur();
	callUpnpDeleteRule(num);
};

return view.extend({
	load: function() {
		return Promise.all([
			callUpnpGetStatus(),
			uci.load('upnpd')
		]);
	},

	poll_status: function(nodes, data) {

		var rules = Array.isArray(data[0].rules) ? data[0].rules : [];

		var rows = rules.map(function(rule) {
			return [
				rule.extport,
				rule.intaddr,
				rule.host_hint || _('Unknown'),
				rule.intport,
				rule.proto,
				rule.descr,
				E('button', {
					'class': 'btn cbi-button-remove',
					'click': L.bind(handleDelRule, this, rule.num)
				}, [ _('Delete') ])
			];
		});

		cbi_update_table(nodes.querySelector('#upnp_status_table'), rows, E('em', _('There are no active port maps.')));

		return;
	},

	render: function(data) {

		var m, s, o;

		m = new form.Map('upnpd', [_('UPnP IGD & PCP/NAT-PMP Service')],
			_('The %s & %s/%s protocols allow clients on the local network to configure port maps/forwards on the router autonomously.', 'The %s (%s = UPnP IGD) & %s (%s = PCP)/%s (%s = NAT-PMP) protocols allow clients on the local network to configure port maps/forwards on the router autonomously.').format('<a href="https://en.wikipedia.org/wiki/Internet_Gateway_Device_Protocol" target="_blank" rel="noreferrer"><abbr title="UPnP Internet Gateway Device (Control Protocol)">UPnP IGD</abbr></a>', '<a href="https://en.wikipedia.org/wiki/Port_Control_Protocol" target="_blank" rel="noreferrer"><abbr title="Port Control Protocol">PCP</abbr></a>', '<a href="https://en.wikipedia.org/wiki/NAT_Port_Mapping_Protocol" target="_blank" rel="noreferrer"><abbr title="NAT Port Mapping Protocol">NAT-PMP</abbr></a>'));

		s = m.section(form.GridSection, '_active_rules');

		s.render = L.bind(function(view, section_id) {
			var table = E('table', { 'class': 'table cbi-section-table', 'id': 'upnp_status_table' }, [
				E('tr', { 'class': 'tr table-titles' }, [
					E('th', { 'class': 'th' }, _('External Port')),
					E('th', { 'class': 'th' }, _('Client Address')),
					E('th', { 'class': 'th' }, _('Host')),
					E('th', { 'class': 'th' }, _('Client Port')),
					E('th', { 'class': 'th' }, _('Protocol')),
					E('th', { 'class': 'th' }, _('Description')),
					E('th', { 'class': 'th cbi-section-actions' }, '')
				])
			]);

			var rules = Array.isArray(data[0].rules) ? data[0].rules : [];

			var rows = rules.map(function(rule) {
				return [
					rule.extport,
					rule.intaddr,
					rule.host_hint || _('Unknown'),
					rule.intport,
					rule.proto,
					rule.descr,
					E('button', {
						'class': 'btn cbi-button-remove',
						'click': L.bind(handleDelRule, this, rule.num)
					}, [ _('Delete') ])
				];
			});

			cbi_update_table(table, rows, E('em', _('There are no active port maps.')));

			return E('div', { 'class': 'cbi-section cbi-tblsection' }, [
					E('h3', _('Active Service Port Maps')), table ]);
		}, o, this);

		s = m.section(form.NamedSection, 'config', 'upnpd', _('Service Settings'));
		s.addremove = false;
		s.tab('general',  _('General Settings'));
		s.tab('advanced', _('Advanced Settings'));

		o = s.taboption('general', form.Flag, 'enabled', _('Enabled'), _('Start autonomous port mapping service'));
		o.rmempty  = false;

		s.taboption('general', form.Flag, 'enable_upnp', _('Enable UPnP IGD protocol')).default = '1';
		s.taboption('general', form.Flag, 'enable_pcp_pmp', _('Enable PCP/NAT-PMP protocols')).default = '1';

		s.taboption('general', form.Flag, 'allow_third_party_maps', _('Allow third-party maps'),
			_('Non-secure mode.'));

		o = s.taboption('general', form.ListValue, 'upnp_igd_compat', _('UPnP IGD compatiblity mode'));
		o.depends('enable_upnp', '1');
		o.value('igdv1',_('IGDv1 (IPv4 only)'));
		o.value('igdv2',_('IGDv2'));

		s.taboption('general', form.Flag, 'ipv6_disable', _('Disable IPv6')).default = '1';

		s.taboption('general', form.Flag, 'force_forwarding', _('Forced forwarding'),
			_('This make the port forwarding force to work even when the router is behind NAT.')).default = '0';

		o = s.taboption('general', form.Value, 'download_kbps', _('Download speed'),
			_('Value in kbit/s, informational only'));
		o.depends('enable_upnp', '1');
		o.rmempty = true;

		o = s.taboption('general', form.Value, 'upload_kbps', _('Upload speed'),
			_('Value in kbit/s, informational only'));
		o.depends('enable_upnp', '1');
		o.rmempty = true;

		o = s.taboption('advanced', form.Value, 'notify_interval', _('Notify interval'), _('A 900s interval will result in %s notifications with the minimum max-age of 1800s', 'A 900s interval will result in %s (%s = SSDP) notifications with the minimum max-age of 1800s').format('<abbr title="Simple Service Discovery Protocol">SSDP</abbr>'));
		o.depends('enable_upnp', '1');
		o.datatype    = 'uinteger';
		o.placeholder = 900;

		o = s.taboption('advanced', form.Value, 'port', _('SOAP/HTTP port'));
		o.depends('enable_upnp', '1');
		o.datatype = 'port';
		o.default  = 5000;

		o = s.taboption('advanced', form.Value, 'presentation_url', _('Presentation URL'));
		o.depends('enable_upnp', '1');
		o.placeholder = 'http://192.168.1.1/';

		o = s.taboption('advanced', form.Value, 'uuid', _('Device UUID'));
		o.depends('enable_upnp', '1');
		o = s.taboption('advanced', form.Value, 'model_number', _('Announced model number'));
		o.depends('enable_upnp', '1');

		o = s.taboption('advanced', form.Value, 'serial_number', _('Announced serial number'));
		o.depends('enable_upnp', '1');

		o = s.taboption('advanced', form.Flag, 'system_uptime', _('Report system instead of service uptime'));
		o.depends('enable_upnp', '1');
		o.default = '1';

		o = s.taboption('advanced', form.Value, 'clean_ruleset_threshold', _('Clean rules threshold (second)'));
		o.depends('enable_upnp', '1');
		o.datatype    = 'uinteger';
		o.placeholder = 20;

		o = s.taboption('advanced', form.Value, 'clean_ruleset_interval', _('Clean rules interval (second)'));
		o.depends('enable_upnp', '1');
		o.datatype    = 'uinteger';
		o.placeholder = 600;

		s.taboption('advanced', form.Flag, 'log_output', _('Enable additional logging'),
			_('Puts extra debugging information into the system log'));

		o = s.taboption('advanced', form.Value, 'upnp_lease_file', _('Service lease file'));
		o.placeholder = '/var/run/miniupnpd.leases';

		s.taboption('advanced', form.Flag, 'use_stun', _('Use STUN'), _('Useful for unrestricted full-cone 1:1 NATs to get the public IPv4 address'));

		o = s.taboption('advanced', form.Value, 'stun_host', _('STUN Host'));
		o.depends('use_stun', '1');
		o.datatype    = 'host';
		o.value('stunserver.stunprotocol.org');
		o.value('stun.syncthing.net');
		o.value('stun.qq.com');
		o.value('stun.miwifi.com');
		o.value('stun.ekiga.net');
		o.value('stun.zoiper.com');
		o.value('stun.gmx.net');
		o.value('stun.counterpath.com');
		o.default  = 'stun.stunprotocol.org';
		o.rmempty  = false;

		o = s.taboption('advanced', form.Value, 'stun_port', _('STUN Port'));
		o.depends('use_stun', '1');
		o.datatype    = 'port';
		o.value('3478');
		o.default  = '3478';
		o.rmempty  = false;

		s = m.section(form.GridSection, 'perm_rule', _('Service ACLs'),
			_('ACLs specify which external ports can be mapped to which client addresses and ports, IPv6 always allowed.'));
		s.sortable  = true;
		s.anonymous = true;
		s.addremove = true;

		s.option(form.Value, 'comment', _('Comment'));

		o = s.option(form.Value, 'ext_ports', _('External Port'));
		o.datatype    = 'portrange';
		o.placeholder = '1-65535';

		o = s.option(form.Value, 'int_addr', _('Client Address'));
		o.datatype    = 'ip4addr';
		o.placeholder = '0.0.0.0/0';

		o = s.option(form.Value, 'int_ports', _('Client Port'));
		o.datatype    = 'portrange';
		o.placeholder = '1-65535';

		o = s.option(form.ListValue, 'action', _('Action'));
		o.value('allow', _('Allow'));
		o.value('deny', _('Deny'));

		return m.render().then(L.bind(function(m, nodes) {
			poll.add(L.bind(function() {
				return Promise.all([
					callUpnpGetStatus()
				]).then(L.bind(this.poll_status, this, nodes));
			}, this), 5);
			return nodes;
		}, this, m));
	}
});
